keps/sig-cli/NNNN-kustomize-components.md

```diff
@@ -0,0 +1,645 @@
+
+## Proposal
+
+Create a new kustomization type called `Component`, and a new kustomization
+field called `components`.
+
+A kustomization that is marked as a `Component` has basically the same
+capabilities as a normal kustomization. The main distinction is that they are
+evaluated **after** the resources of the parent kustomization (overlay or
+component) have been accumulated, and **on top** of them. This means that:
+
+* A component with transformers can transform the resources that an overlay has
+  previously specified in the `resources` field. Components with patches do not
+  have to include the target resource in their `resources` field.
+* Multiple components can extend and transform the same set of resources
+  **sequentially**. This is in contrast to overlays, which cannot alter the same
+  base resources, because they clone them and extend them **in parallel**.
+
+In order to create a component, a user can add the following to its
+`kustomization.yaml`:
+
+```yaml
+apiVersion: kustomize.config.k8s.io/v1alpha1
+kind: Component
+```
+
+In order to use a component, the user can refer to it in their
+`kustomization.yaml` via the `components` field:
+
+```yaml:
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+
+resources:
+  - ../../base
+  - resource1
+  - resource2
+
+components:
+  - ../components/component1
+  - ../components/component2
+```
+
+Note that a component cannot be added to the `resources:` list, and a
+resource/`Kustomization` cannot be added to the `components:` list.
+
+### User Story
+
+Suppose that a company has written a very simple Web application:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: example
+spec:
+  template:
+    spec:
+      containers:
+      - name: example
+        image: example:1.0
+```
+
+They may want to deploy a **community** edition of this application as SaaS, so
+they need to add support for persistence (e.g. an external database), and bot
+detection (e.g. Google reCAPTCHA).
+
+At some point, they have attracted **enterprise** customers who want to deploy
+it on-premises, so they add LDAP support, and disable Google reCAPTCHA. At the
+same time, the **devs** need to be able to test parts of the application, so
+they want to deploy it with some features enabled and others not.
+
+Here's a matrix with the deployments of this application and the features
+enabled for each one:
+
+|            | External DB        | LDAP               | reCAPTCHA          |
+|------------|:------------------:|:------------------:|:------------------:|
+| Community  | :heavy_check_mark: |                    | :heavy_check_mark: |
+| Enterprise | :heavy_check_mark: | :heavy_check_mark: |                    |
+| Dev        | :white_check_mark: | :white_check_mark: | :white_check_mark: |
+
+(:heavy_check_mark:: enabled, :white_check_mark:: optional)
+
+So, we want to make it easy to deploy this application in any of the above
```

Comment:
Here we say "we", but we probably need "they" to be inline with the rest of the section.

Also see the previous review comment.

Reply:
I've updated the wording here, give it a look.